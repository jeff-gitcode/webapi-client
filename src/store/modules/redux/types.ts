export enum ActionsTypes {
  addToCart = "ADD_TO_CART",

  fetchProducts = "FETCH_PRODUCTS",
  fetchProductsSuccess = "FETCH_PRODUCTS_SUCCCESS",
  fetchProductsFailure = "FETCH_PRODUCTS_FAILURE",

  addShipping = "ADD_SHIPPING",
  substractShipping = "SUB_SHIPPING",

  shippingCost = "SHIPPINGCOST",
  shippingCostSuccess = "SHIPPINGCOST_SUCCESS",
  shippingCostFailure = "SHIPPINGCOST_FAILURE",

  changeCurrency = "CHANGE_CURRENCY",

  checkout = "CHECKOUT",
  checkoutSuccess = "CHECKOUT_SUCCESS",
  checkoutFailure = "CHECKOUT_FAILURE",

  resetShop = "RESET_SHOP",

  removeItem = "REMOVE_ITEM",

  subtractQuantity = "SUB_QUANTITY",
  addQuantity = "ADD_QUANTITY",
}

export interface ListResponse<T> {
  data: T[];
}

export interface ICheckOut {
  items: IProduct[];
  total: number;
}

export interface IShipping {
  total: number;
}

export interface IProduct {
  id: number;
  title: string;
  price: number;
  quantity?: number;
}

export interface IState {
  items: IProduct[];
  addedItems: IProduct[];
  shipping: number;
  total: number;
  complete: boolean;
  error: string;
}
