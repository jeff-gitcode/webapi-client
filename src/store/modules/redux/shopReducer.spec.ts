import reducer from "./shopReducer";
import { ActionsTypes } from "./types";
import {
  addQuantity,
  addToCart,
  changeCurrency,
  checkoutFailure,
  checkoutSuccess,
  fetchProducts,
  fetchProductsFailure,
  fetchProductsSuccess,
  removeItem,
  resetShop,
  shippingCostFailure,
  shippingCostSuccess,
  substractShipping,
  subtractQuantity,
} from "./actions";

describe("reducers/fetch products", () => {
  const emptyAction = { type: "", payload: {} };
  let currentState: any = null;

  beforeAll(() => {
    currentState = reducer(undefined, emptyAction);
  });

  afterAll(() => {
    currentState = null;
  });

  it("should return the initial state", () => {
    expect(reducer(currentState, emptyAction)).toMatchSnapshot();
  });

  it(`should handle ${ActionsTypes.resetShop}`, () => {
    currentState = reducer(currentState, resetShop());
    expect(currentState).toMatchSnapshot();
  });

  it(`should handle ${ActionsTypes.fetchProducts}`, () => {
    currentState = reducer(currentState, fetchProducts());
    expect(currentState).toMatchSnapshot();
  });

  it(`should handle ${ActionsTypes.fetchProductsSuccess}`, () => {
    const payload = {
      items: [
        {
          id: 1,
          title: "Casaco maneiro",
          price: 290.9,
          quantity: 0,
        },
      ],
    };

    currentState = reducer(currentState, fetchProductsSuccess(payload));
    expect(currentState).toMatchSnapshot();
  });

  it(`should handle ${ActionsTypes.addToCart}`, () => {
    const payload = { id: 1 };

    currentState = reducer(currentState, addToCart(payload));
    expect(currentState).toMatchSnapshot();
  });

  it(`should handle ${ActionsTypes.addQuantity}`, () => {
    const payload = { id: 1 };

    currentState = reducer(currentState, addQuantity(payload));
    expect(currentState).toMatchSnapshot();
  });

  it(`should handle ${ActionsTypes.subtractQuantity}`, () => {
    const payload = { id: 1 };

    currentState = reducer(currentState, subtractQuantity(payload));
    expect(currentState).toMatchSnapshot();
  });

  it(`should handle ${ActionsTypes.shippingCostSuccess}`, () => {
    const payload = { shipping: 10 };

    currentState = reducer(currentState, shippingCostSuccess(payload));
    expect(currentState).toMatchSnapshot();
  });

  it(`should handle ${ActionsTypes.changeCurrency}`, () => {
    const payload = { currency: "2" };

    currentState = reducer(currentState, changeCurrency(payload));
    expect(currentState).toMatchSnapshot();
  });

  it(`should handle ${ActionsTypes.substractShipping}`, () => {
    currentState = reducer(currentState, substractShipping());
    expect(currentState).toMatchSnapshot();
  });

  it(`should handle ${ActionsTypes.removeItem}`, () => {
    const payload = { id: 1 };

    currentState = reducer(currentState, removeItem(payload));
    expect(currentState).toMatchSnapshot();
  });

  it(`should handle ${ActionsTypes.checkoutSuccess}`, () => {
    const payload = { complete: true };

    currentState = reducer(currentState, checkoutSuccess(payload));
    expect(currentState).toMatchSnapshot();
  });

  it(`should handle ${ActionsTypes.fetchProductsFailure}`, () => {
    const payload = { error: "error" };

    currentState = reducer(currentState, fetchProductsFailure(payload));
    expect(currentState).toMatchSnapshot();
  });

  it(`should handle ${ActionsTypes.shippingCostFailure}`, () => {
    const payload = { error: "error" };

    currentState = reducer(currentState, shippingCostFailure(payload));
    expect(currentState).toMatchSnapshot();
  });

  it(`should handle ${ActionsTypes.checkoutFailure}`, () => {
    const payload = { error: "error" };

    currentState = reducer(currentState, checkoutFailure(payload));
    expect(currentState).toMatchSnapshot();
  });
});
