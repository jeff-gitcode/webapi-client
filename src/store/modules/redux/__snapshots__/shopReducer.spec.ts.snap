// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers/fetch products should handle ADD_QUANTITY 1`] = `
Object {
  "addedItems": Array [
    Object {
      "id": 1,
      "price": 290.9,
      "quantity": 2,
      "title": "Casaco maneiro",
    },
  ],
  "complete": false,
  "error": "",
  "items": Array [
    Object {
      "id": 1,
      "price": 290.9,
      "quantity": 2,
      "title": "Casaco maneiro",
    },
  ],
  "shipping": 0,
  "total": 581.8,
}
`;

exports[`reducers/fetch products should handle ADD_TO_CART 1`] = `
Object {
  "addedItems": Array [
    Object {
      "id": 1,
      "price": 290.9,
      "quantity": 1,
      "title": "Casaco maneiro",
    },
  ],
  "complete": false,
  "error": "",
  "items": Array [
    Object {
      "id": 1,
      "price": 290.9,
      "quantity": 1,
      "title": "Casaco maneiro",
    },
  ],
  "shipping": 0,
  "total": 290.9,
}
`;

exports[`reducers/fetch products should handle CHANGE_CURRENCY 1`] = `
Object {
  "addedItems": Array [
    Object {
      "id": 1,
      "price": 581.8,
      "quantity": 1,
      "title": "Casaco maneiro",
    },
  ],
  "complete": false,
  "error": "",
  "items": Array [
    Object {
      "id": 1,
      "price": 290.9,
      "quantity": 2,
      "title": "Casaco maneiro",
    },
  ],
  "shipping": 10,
  "total": 581.8,
}
`;

exports[`reducers/fetch products should handle CHECKOUT_FAILURE 1`] = `
Object {
  "addedItems": Array [],
  "complete": true,
  "error": "checkoutFailure: error",
  "items": Array [],
  "shipping": 0,
  "total": -10,
}
`;

exports[`reducers/fetch products should handle CHECKOUT_SUCCESS 1`] = `
Object {
  "addedItems": Array [],
  "complete": true,
  "error": "",
  "items": Array [
    Object {
      "id": 1,
      "price": 290.9,
      "quantity": 2,
      "title": "Casaco maneiro",
    },
  ],
  "shipping": 0,
  "total": -10,
}
`;

exports[`reducers/fetch products should handle FETCH_PRODUCTS 1`] = `
Object {
  "addedItems": Array [],
  "complete": false,
  "error": "",
  "items": Array [],
  "shipping": 0,
  "total": 0,
}
`;

exports[`reducers/fetch products should handle FETCH_PRODUCTS_FAILURE 1`] = `
Object {
  "addedItems": Array [],
  "complete": true,
  "error": "fetchProductsFailure: error",
  "items": Array [],
  "shipping": 0,
  "total": -10,
}
`;

exports[`reducers/fetch products should handle FETCH_PRODUCTS_SUCCCESS 1`] = `
Object {
  "addedItems": Array [],
  "complete": false,
  "error": "",
  "items": Array [
    Object {
      "id": 1,
      "price": 290.9,
      "quantity": 0,
      "title": "Casaco maneiro",
    },
  ],
  "shipping": 0,
  "total": 0,
}
`;

exports[`reducers/fetch products should handle REMOVE_ITEM 1`] = `
Object {
  "addedItems": Array [],
  "complete": false,
  "error": "",
  "items": Array [
    Object {
      "id": 1,
      "price": 290.9,
      "quantity": 2,
      "title": "Casaco maneiro",
    },
  ],
  "shipping": 0,
  "total": -10,
}
`;

exports[`reducers/fetch products should handle RESET_SHOP 1`] = `
Object {
  "addedItems": Array [],
  "complete": false,
  "error": "",
  "items": Array [],
  "shipping": 0,
  "total": 0,
}
`;

exports[`reducers/fetch products should handle SHIPPINGCOST_FAILURE 1`] = `
Object {
  "addedItems": Array [],
  "complete": true,
  "error": "shippingCostFailure: error",
  "items": Array [],
  "shipping": 0,
  "total": -10,
}
`;

exports[`reducers/fetch products should handle SHIPPINGCOST_SUCCESS 1`] = `
Object {
  "addedItems": Array [
    Object {
      "id": 1,
      "price": 290.9,
      "quantity": 1,
      "title": "Casaco maneiro",
    },
  ],
  "complete": false,
  "error": "",
  "items": Array [
    Object {
      "id": 1,
      "price": 290.9,
      "quantity": 2,
      "title": "Casaco maneiro",
    },
  ],
  "shipping": 10,
  "total": 300.9,
}
`;

exports[`reducers/fetch products should handle SUB_QUANTITY 1`] = `
Object {
  "addedItems": Array [
    Object {
      "id": 1,
      "price": 290.9,
      "quantity": 1,
      "title": "Casaco maneiro",
    },
  ],
  "complete": false,
  "error": "",
  "items": Array [
    Object {
      "id": 1,
      "price": 290.9,
      "quantity": 2,
      "title": "Casaco maneiro",
    },
  ],
  "shipping": 0,
  "total": 290.9,
}
`;

exports[`reducers/fetch products should handle SUB_SHIPPING 1`] = `
Object {
  "addedItems": Array [
    Object {
      "id": 1,
      "price": 581.8,
      "quantity": 1,
      "title": "Casaco maneiro",
    },
  ],
  "complete": false,
  "error": "",
  "items": Array [
    Object {
      "id": 1,
      "price": 290.9,
      "quantity": 2,
      "title": "Casaco maneiro",
    },
  ],
  "shipping": 0,
  "total": 571.8,
}
`;

exports[`reducers/fetch products should return the initial state 1`] = `
Object {
  "addedItems": Array [],
  "complete": false,
  "error": "",
  "items": Array [],
  "shipping": 0,
  "total": 0,
}
`;
